processor/deltatocumulativeprocessor/internal/data/add.go

```diff
@@ -65,25 +65,95 @@ func (dp Histogram) Add(in Histogram) Histogram {
 	return dp
 }
 
+type bounds struct {
+	lower int32
+	upper int32
+}
+
+// with is an accessory for Merge() to calculate ideal combined scale.
+func (b bounds) with(o bounds) bounds {
+	if o.empty() {
+		return b
+	}
+	if b.empty() {
+		return o
+	}
+	return bounds{
+		lower: min(b.lower, o.lower),
+		upper: max(b.upper, o.upper),
+	}
+}
+
+// empty indicates whether there are any values in a bounds.
+func (b bounds) empty() bool {
+	return b == bounds{}
+}
+
+// boundsAtScale is an accessory for Add() to calculate ideal combined scale.
+func (dp ExpHistogram) boundsAtScale(b expo.Buckets, scale int32) bounds {
+	if b.BucketCounts().Len() == 0 {
+		return bounds{}
+	}
+	shift := dp.Scale() - scale
+	a := expo.Abs(b)
+	return bounds{
+		lower: int32(a.Lower()) >> shift,
+		upper: int32(a.Upper()) >> shift,
+	}
+}
+
+// downscaleNeeded computes how much downscaling is needed by shifting the
+// upper and lower bounds until they are separated by no more than size.
+func downscaleNeeded(b bounds, size int) int32 {
+	var change int32
+
+	for b.upper-b.lower > int32(size) {
+		b.upper >>= 1
+		b.lower >>= 1
+		change++
+	}
+	return change
+}
+
 func (dp ExpHistogram) Add(in ExpHistogram) ExpHistogram {
 	type H = ExpHistogram
 
-	if dp.Scale() != in.Scale() {
-		hi, lo := expo.HiLo(dp, in, H.Scale)
-		from, to := expo.Scale(hi.Scale()), expo.Scale(lo.Scale())
-		expo.Downscale(hi.Positive(), from, to)
-		expo.Downscale(hi.Negative(), from, to)
-		hi.SetScale(lo.Scale())
-	}
+	minScale := min(dp.Scale(), in.Scale())
+
+	// logic is adapted from lightstep's algorithm for enforcing max buckets:
+	// https://github.com/lightstep/go-expohisto/blob/4375bf4ef2858552204edb8b4572330c94a4a755/structure/exponential.go#L542
+	// first, calculate the highest and lowest indices for each bucket, given the candidate min scale.
+	// then, calculate how much downscaling is needed to fit the merged range within max bucket count.
+	// finally, perform the actual downscaling.
+	posBounds := dp.boundsAtScale(dp.Positive(), minScale)
+	posBounds = posBounds.with(in.boundsAtScale(in.Positive(), minScale))
+
+	negBounds := dp.boundsAtScale(dp.Negative(), minScale)
+	negBounds = negBounds.with(in.boundsAtScale(in.Negative(), minScale))
+
+	minScale = min(
+		minScale-downscaleNeeded(posBounds, dp.MaxSize),
+		minScale-downscaleNeeded(negBounds, dp.MaxSize),
+	)
```

Comment:
can we move all of this logic into the `expo` package? This is generally useful and fits that packages purpose better.

maybe:
```go
package expo

// Limit returns the Scale a and b need to be downscaled to so that merging does
// not exceed the given max bucket length
func Limit(a, b DataPoint, max int) Scale {}
```

Reply:
Sounds good.